hamilton/experimental/h_spark.py

```diff
@@ -149,3 +149,283 @@ def build_result(
+
+        :param kwargs: the inputs to the function.
+        :return: tuple of the dataframe and the remaining non-dataframe kwargs.
+        """
+        df = None
+        actual_kwargs = {}
+        for kwarg_key, kwarg_value in kwargs.items():
+            if isinstance(kwarg_value, DataFrame):
+                if df is None:
+                    df = kwarg_value
+            else:
+                actual_kwargs[kwarg_key] = kwarg_value
+        return df, actual_kwargs
+
+    def execute_node(self, node: node.Node, kwargs: Dict[str, Any]) -> Any:
+        """Given a node to execute, process it and apply a UDF if applicable.
+
+        :param node: the node we're processing.
+        :param kwargs: the inputs to the function.
+        :return: the result of the function.
+        """
+        self.call_count += 1
+        logger.debug("%s, %s", self.call_count, self.df_object)
+        # get dataframe object out of kwargs
+        df, actual_kwargs = self._inspect_kwargs(kwargs)
+        if df is None:  # there were no dataframes passed in. So regular function call.
+            return node.callable(**actual_kwargs)
+        if self.df_object is None:
+            self.df_object = df  # this is done only once.
+            self.original_schema = list(df.columns)
+        logger.debug("%s, %s", self.call_count, self.df_object)
+        logger.debug("%s, Before, %s", node.name, self.df_object.columns)
+        schema_length = len(df.schema)
+        df = self._lambda_udf(self.df_object, node, node.callable, actual_kwargs)
+        assert node.name in df.columns, f"Error {node.name} not in {df.columns}"
+        delta = len(df.schema) - schema_length
+        if delta == 0:
+            raise ValueError(
+                f"UDF {node.name} did not add any columns to the dataframe. "
+                f"Does it already exist in the dataframe?"
+            )
+        self.df_object = df
+        logger.debug("%s, After, %s", node.name, df.columns)
+        return df
+
+    def build_result(self, **outputs: Dict[str, Any]) -> DataFrame:
+        """Builds the result and brings it back to this running process.
+
+        :param outputs: the dictionary of key -> Union[ray object reference | value]
+        :return: The type of object returned by self.result_builder.
+        """
+        df: DataFrame = self.df_object
+        output_schema = self.original_schema
+        # what's in the dataframe:
+        for output_name, output_value in outputs.items():
+            if output_name not in output_schema:
+                output_schema.append(output_name)
+            if output_name in df.columns:
+                continue
+            else:
+                df = df.withColumn(output_name, lit(output_value))
+        # original schema + new columns should be the order.
+        # if someone requests a column that is in the original schema we won't duplicate it.
+        result = df.select(*[column(col_name) for col_name in output_schema])
```

Comment:
nit here, but you can build select expr here rather than doing df.withColumn. This will affect performance in wide dataframes.

Reply:
code snippet please? :)

Reply:
Sure, 

```python
def some_fun(df):
    outputs={'col1': 23.0, 'col2': 45.0, 'col3': 67.0}
    df_columns=set(df.columns)
    select_exprs=[]
    for output_name, output_value in outputs.items():
        if output_name in df_columns:
            select_exprs.append(column(output_name))
        else:
            select_exprs.append(lit(output_value).alias(output_name))
    return df.select(*select_exprs)
```

Reply:
Ahh I remember last time you showed us this. IIRC the issue isn't the actual computation, its the fact that the plan gets big and it takes longer to optimize? https://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html#withColumn(colName:String,col:org.apache.spark.sql.Column):org.apache.spark.sql.DataFrame

Reply:
Unfortunately python doesn't have `withColumns` :/ Just java/scala.

Reply:
IIRC, it leads to a df copy, and that slows stuff down.

Reply:
from the docs... 
2.0.0

Note
`this method introduces a projection internally. Therefore, calling it multiple times, for instance, via loops in order to add multiple columns can generate big plans which can cause performance issues and even StackOverflowException. To avoid this, use select with the multiple columns at once.`

Reply:
Yeah, the problem is the plan right? Not the actual computation that gets run? E.G. executing the plan gets ugly (its kind of an n^2 set of parameters...)

Reply:
Hold up. The `with_column` here is only called for a minority case. So that's likely never going to be a bottleneck.

What this does bring up though, is whether the internal function to call the UDFs will be. Since it does `with_column` for each UDF... 🤔

Reply:
Yeah I'm not going to refactor this now -- but yeah it would be a bit of surgery to rewrite things to return udfs, and then only in build_result do the final select expression. I am currently using the dataframe object to tell me what columns should exist in it -- so would need to replace that with an object, etc...

Reply:
This will get slow if there's  1000 columns -- but I think its not the end of the world right now, and not worth blocking.